File: ServerTest/Context/Test/src/main/java/org/elkoserver/server/context/test/Counter.java
package org.elkoserver.server.context.test;

import org.elkoserver.foundation.json.JSONMethod;
import org.elkoserver.foundation.json.MessageHandlerException;
import org.elkoserver.json.EncodeControl;
import org.elkoserver.json.JSONLiteral;
import org.elkoserver.server.context.GeneralMod;
import org.elkoserver.server.context.Mod;
import org.elkoserver.server.context.User;

import static org.elkoserver.json.JSONLiteralFactory.targetVerb;
import static org.elkoserver.json.JSONLiteralFactory.type;

/**
 * Mod to enable an object to function as a simple counter, mainly for testing
 * control over persistence.
 */
public class Counter extends Mod implements GeneralMod {
    /** Current count. */
    private int myCount;

    /**
     * JSON-driven constructor.
     */
    @JSONMethod({ "count" })
    public Counter(int count) {
        myCount = count;
    }

    /**
     * Encode this mod for transmission or persistence.
     *
     * @param control  Encode control determining what flavor of encoding
     *    should be done.
     *
     * @return a JSON literal representing this mod.
     */
    public JSONLiteral encode(EncodeControl control) {
        JSONLiteral result = type("counter", control);
        result.addParameter("count", myCount);
        result.finish();
        return result;
    }

    /**
     * Message handler for the 'inc' message.
     *
     * This message requests the counter to increment.
     *
     * <u>recv</u>: <code> { to:<i>REF</i>, op:"inc" } </code><br>
     *
     * <u>send</u>: <code> { to:<i>REF</i>, op:"set",
     *                     from:<i>REF</i>, count:<i>int</i> } </code>
     *
     * @param from  The user requesting the census.
     *
     * @throws MessageHandlerException if 'from' is not in the same context as
     *    this mod, or if this mod is attached to a user and 'from' is not that
     *    user.
     */
    @JSONMethod
    public void inc(User from) throws MessageHandlerException {
        ensureSameContext(from);
        ++myCount;
        markAsChanged();
        JSONLiteral announce = targetVerb(object(), "set");
        announce.addParameter("count", myCount);
        announce.finish();
        context().send(announce);
    }
}

File: ServerTest/Context/Test/src/main/kotlin/org/elkoserver/server/context/test/Counter.kt
package org.elkoserver.server.context.test

import org.elkoserver.foundation.json.JsonMethod
import org.elkoserver.foundation.json.MessageHandlerException
import org.elkoserver.json.EncodeControl
import org.elkoserver.json.JsonLiteral
import org.elkoserver.json.JsonLiteralFactory
import org.elkoserver.server.context.model.GeneralMod
import org.elkoserver.server.context.model.Mod
import org.elkoserver.server.context.model.User

/**
 * Mod to enable an object to function as a simple counter, mainly for testing
 * control over persistence.
 */
class Counter @JsonMethod("count") constructor(private var myCount: Int) : Mod(), GeneralMod {

    /**
     * Encode this mod for transmission or persistence.
     *
     * @param control  Encode control determining what flavor of encoding
     * should be done.
     *
     * @return a JSON literal representing this mod.
     */
    override fun encode(control: EncodeControl): JsonLiteral =
            JsonLiteralFactory.type("counter", control).apply {
                addParameter("count", myCount)
                finish()
            }

    /**
     * Message handler for the 'inc' message.
     *
     * This message requests the counter to increment.
     *
     * <u>recv</u>: ` { to:*REF*, op:"inc" } `<br></br>
     *
     * <u>send</u>: ` { to:*REF*, op:"set",
     * from:*REF*, count:*int* } `
     *
     * @param from  The user requesting the census.
     *
     * @throws MessageHandlerException if 'from' is not in the same context as
     * this mod, or if this mod is attached to a user and 'from' is not that
     * user.
     */
    @JsonMethod
    fun inc(from: User) {
        ensureSameContext(from)
        ++myCount
        markAsChanged()
        val announce = JsonLiteralFactory.targetVerb(`object`(), "set").apply {
            addParameter("count", myCount)
            finish()
        }
        context().send(announce)
    }
}
